waspc/data/Generator/templates/react-app/src/queries/core.js

```diff
@@ -1,9 +1,16 @@
 import { callOperation } from '../operations'
-import { addResourcesUsedByQuery } from '../operations/resources'
+import {
+  addResourcesUsedByQuery,
+  getPendingUpdatesForQuery
+} from '../operations/resources'
 
 export function createQuery(queryRoute, entitiesUsed) {
-  function query(args) {
-    return callOperation(queryRoute, args)
+  async function query(queryKey, queryArgs) {
+    const serverResult = await callOperation(queryRoute, queryArgs)
+    return getPendingUpdatesForQuery(queryKey).reduce(
+      (result, pendingUpdate) => pendingUpdate(result), 
+      serverResult,
+    )
```

Comment:
This is where the magic happens. We get a list of all unresolved optimistic updates and apply all of them to the query's result.

Since we do it for all queries, it automatically solves the problems we had with React Query's internal invalidations. In other words, our logic does not need to know or care about how the queries are invalidated (as opposed to the counter approach).

Reply:
This is a really neat idea to solve the counter problems! As you noted, the only real risk here is if their update function does something super whacky.

Also, we will want to check for error conditions for pending operations. I updated an action to fail 50% of the time, and hit this error:
<img width="1520" alt="Screenshot 2022-07-19 134652" src="https://user-images.githubusercontent.com/523636/179820310-59d6864b-4af5-4ee7-80e2-85026cde7906.png">

Reply:
Cool! Does the order of these updates matter in some way?
For example, if user makes 3 actions, A1, A2 and A3, in that order, all affecting the same query, and we get result of A2 first, then A3 and then A1, what happens then? Hm I guess this is up to them, to define this as part of business logic? Because if they wanted them to happen in sequence, they should have done it so themselves? Ok I am asking a bit vague question because I am missing context to ask a better question, but I hope you might be able to get more out if that I do @sodic .

Reply:
@shayneczyzewski This was an issue with the original implementation (i.e., not related to fixing flickering UI). Anyway, I fixed it in this commit: https://github.com/wasp-lang/wasp/pull/672/commits/b48a7ad7343fc1563294aaa3da21b80de18bc111. Dumb mistake :sweat_smile:

Reply:
@sodic just to confirm, you said offline to me that the order of actions is basically not a problem, right? If so then all good!